File: kyulee/programmers-2/행렬테두리회전하기.java
import java.util.*;

public class 행렬테두리회전하기 {
    class Solution {
        public int[] solution(int rows, int columns, int[][] queries) {
            // 주어진 행렬 생성 및 초기화
            int[][] matrix = new int[rows][columns];
            int num = 1;
            for (int i = 0; i < rows; i++) {
                for (int j = 0; j < columns; j++) {
                    matrix[i][j] = num++;
                }
            }

            // 결과를 담을 배열
            int[] result = new int[queries.length];

            // 각 회전에 대해 순회
            for (int i = 0; i < queries.length; i++) {
                // 회전을 적용하고, 회전된 테두리의 최솟값을 찾아 결과 배열에 추가
                result[i] = rotateAndFindMin(matrix, queries[i]);
            }

            return result;
        }

        // 주어진 행렬에서 테두리를 시계방향으로 회전하고, 회전된 테두리 중 최솟값을 반환하는 함수
        private int rotateAndFindMin(int[][] matrix, int[] query) {
            int x1 = query[0] - 1, y1 = query[1] - 1, x2 = query[2] - 1, y2 = query[3] - 1;

            // 상단 테두리
            int prev = matrix[x1][y1];
            int min = prev;
            for (int i = y1 + 1; i <= y2; i++) {
                min = Math.min(min, prev);
                int temp = matrix[x1][i];
                matrix[x1][i] = prev;
                prev = temp;
            }

            // 우측 테두리
            for (int i = x1 + 1; i <= x2; i++) {
                min = Math.min(min, prev);
                int temp = matrix[i][y2];
                matrix[i][y2] = prev;
                prev = temp;
            }

            // 하단 테두리
            for (int i = y2 - 1; i >= y1; i--) {
                min = Math.min(min, prev);
                int temp = matrix[x2][i];
                matrix[x2][i] = prev;
                prev = temp;
            }

            // 좌측 테두리
            for (int i = x2 - 1; i >= x1; i--) {
                min = Math.min(min, prev);
                int temp = matrix[i][y1];
                matrix[i][y1] = prev;
                prev = temp;
            }

            return min;
        }
    }
}

File: soyeong/programmers-2/행렬-테두리-회전하기.py
def solution(rows, columns, queries):
    # 값이 들어간 다차원배열 생성
    arr = [[(i) * columns + (j + 1) for j in range(columns)] for i in range(rows)]
    result = []
    for x1, y1, x2, y2 in queries:
        result.append(change(x1 - 1, y1 - 1, x2 - 1, y2 - 1, arr))
    return result

def change(x1, y1, x2, y2, arr):
    # 가장 큰 수부터 비교하며 최소 숫자를 찾아가야하므로 이 조건에서 들어올 수 있는 가장 큰 수를 할당
    # rows 최대 100 columns 최대 100 이니 될수 있는 가장 큰 수 10000
    min_value = 10000

    # 시작 좌표 저장
    temp = arr[x1][y1]

    # 아래 4가지 방향으로 이동하며 하나씩 당겨 값을 저장
    # 가장 첫번째 값은 temp에서 저장하고 있으므로 첫번째 값이 덮어지는 것은 괜찮음

    # 왼 > 오
    for k in range(x1, x2):
        arr[k][y1] = arr[k + 1][y1]
        min_value = min(min_value, arr[k + 1][y1])  # 이동 중 가장 작은 값을 비교해서 저장
    # 위 > 아래
    for k in range(y1, y2):
        arr[x2][k] = arr[x2][k + 1]
        min_value = min(min_value, arr[x2][k + 1])
    # 오 > 왼
    for k in range(x2, x1, -1):
        arr[k][y2] = arr[k - 1][y2]
        min_value = min(min_value, arr[k - 1][y2])
    # 아래 > 위
    for k in range(y2, y1 + 1, -1):
        arr[x1][k] = arr[x1][k - 1]
        min_value = min(min_value, arr[x1][k - 1])
    # 기록했던 처음값을 처음자리 아래에 업데이트하여 전체 회전 적용 완료
    arr[x1][y1 + 1] = temp

    # 최초 값까지 포함하여 비교해 이동한 값 중 가장 작은 값 저장
    min_value = min(min_value, temp)
    return min_value